packages/nouns-contracts/contracts/client-incentives/Rewards.sol

```diff
@@ -320,11 +325,17 @@ contract Rewards is
         require(lastProposalId >= t.nextProposalIdToReward, 'bad lastProposalId');
         require(isSortedAndNoDuplicates(votingClientIds), 'must be sorted & unique');
 
+        uint16 proposalEligibilityQuorumBps_ = $.params.proposalEligibilityQuorumBps;
+
         NounsDAOTypes.ProposalForRewards[] memory proposals = nounsDAO.proposalDataForRewards(
             t.nextProposalIdToReward,
             lastProposalId,
+            proposalEligibilityQuorumBps_,
+            true,
             votingClientIds
         );
+        require(proposals.length > 0, 'at least one eligible proposal');
+        t.numEligibleProposals = proposals.length;
         $.nextProposalIdToReward = lastProposalId + 1;
```

Comment:
Should be `$.nextProposalIdToReward = t.lastProposal + 1` (`t.lastProposal` setting needs to go up accordingly) otherwise there can be griefing.

Reply:
You're right, documenting the potential griefing with more detail:
because the check that voting is over is done only on the filtered proposals returned from `proposalDataForRewards`, `lastProposalId` has only limited to be `<= nounsDAO.proposalCount()`.
Someone can provide a `lastProposalId` of a proposal that was just created and it would be processed as completed and rewards would be skipped for that proposal.

Reply:
Thinking about it more, another potential issue:
The check that voting has finished is now done only on proposals returned from `proposalDataForRewards`.
Could there be a proposal earlier than `t.lastProposal` that has not yet finished voting, and has been filtered out in `proposalDataForRewards`?
If objection period is turned on for a proposal, then its voting period could end *after* a later proposal.
But, for objection period to be turn on it means that the proposal got at least enough votes to pass, using the DAO's dynamic quorum. So I think in practice there's low risk here.

Another potential situation in which an earlier proposal can finish voting after a later one is in case the DAO decreases the votingPeriod parameter.
In this case the code here could filter out the earlier proposal, even though it may get more votes in the future.

For these reasons, I think it's worthwhile to check that all proposals up to `nextProposalIdToReward` have finished voting.
With the current code design, this means adding this check to the loop in `proposalDataForRewards`:
```
            // make sure proposal finished voting
            uint endBlock = max(proposals[i].endBlock, proposals[i].objectionPeriodEndBlock);
            require(block.number > endBlock, 'all proposals must be done with voting');
```

@eladmallel @dmitriia please lmk what you think

Reply:
> You're right, documenting the potential griefing with more detail: because the check that voting is over is done only on the filtered proposals returned from `proposalDataForRewards`, `lastProposalId` has only limited to be `<= nounsDAO.proposalCount()`. Someone can provide a `lastProposalId` of a proposal that was just created and it would be processed as completed and rewards would be skipped for that proposal.

I think this is correct, any proposal in a early stage, having votes below the threshold, can be attacked this way. Also, it can be an indirect/by-product impact with other proposal's beneficiaries just wanting to receive their rewards, not intending to grief the unfinished ones in the first place.

Reply:
Since only `against` votes can be casted during the objection period (so it cannot be not eligible before and eligible after just by `forVotes` accumulation), the only viable scenarios seems to be `votingPeriod` or `proposalEligibilityQuorumBps` decreases. If `votingPeriod` was decreased between two proposals then a proposal that still can go beyond the eligibility threshold can be filtered out if the next one is already concluded. If `proposalEligibilityQuorumBps` was decreased then running the rewards allocation just before that decrease can deprive a proposal based on the old rules. This is, however, a kind of usual race condition on configuration changes.

> For these reasons, I think it's worthwhile to check that all proposals up to `nextProposalIdToReward` have finished voting. With the current code design, this means adding this check to the loop in `proposalDataForRewards`:
> 
> ```
>             // make sure proposal finished voting
>             uint endBlock = max(proposals[i].endBlock, proposals[i].objectionPeriodEndBlock);
>             require(block.number > endBlock, 'all proposals must be done with voting');
> ```
> 
> @eladmallel @dmitriia please lmk what you think

I support this. I think it has to go after `cancel` check as `Canceled` is a final state and it will not matter how far is `endBlock`.

Also, it will cover the `votingPeriod` change, but not `proposalEligibilityQuorumBps` change. It looks to be fine as the latter is a kind of expected behavior.